services/versioncache/versioncache.go

```diff
@@ -0,0 +1,303 @@
+	results := &prepareResults{CheckAfter: minCheckInterval}
+
+	if err := s.db.InTransaction(func(tx *reform.TX) error {
+		filter := models.FindServicesSoftwareVersionsFilter{Limit: pointer.ToInt(1)}
+		servicesVersions, err := models.FindServicesSoftwareVersions(tx.Querier, filter)
+		if err != nil {
+			return err
+		}
+		if len(servicesVersions) == 0 {
+			results.CheckAfter = serviceCheckInterval
+
+			return nil
+		}
+		if servicesVersions[0].NextCheckAt.After(time.Now()) {
+			results.CheckAfter = time.Until(servicesVersions[0].NextCheckAt)
+			if results.CheckAfter < minCheckInterval {
+				results.CheckAfter = minCheckInterval
+			}
+
+			return nil
+		}
+
+		results.NeedsUpdate = true
+		results.ServiceID = servicesVersions[0].ServiceID
+
+		service, err := models.FindServiceByID(tx.Querier, servicesVersions[0].ServiceID)
+		if err != nil {
+			return err
+		}
+		if service.ServiceType != models.MySQLServiceType {
+			return nil
+		}
+
+		pmmAgents, err := models.FindPMMAgentsForService(tx.Querier, servicesVersions[0].ServiceID)
+		if err != nil {
+			return err
+		}
+		if len(pmmAgents) == 0 {
+			return errors.Errorf("pmmAgent not found for service")
+		}
+		results.PMMAgentID = pmmAgents[0].AgentID
+
+		// shift the next check time for this service, so, in case of versions fetch error,
+		// it will not loop in trying, but will continue with other services.
+		nextCheckAt := time.Now().UTC().Add(serviceCheckInterval)
+		if _, err := models.UpdateServiceSoftwareVersions(tx.Querier, servicesVersions[0].ServiceID,
+			models.UpdateServiceSoftwareVersionsParams{NextCheckAt: &nextCheckAt},
+		); err != nil {
+			return err
+		}
+
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	return results, nil
+}
+
+func softwareName(s agents.Software) (models.SoftwareName, error) {
+	var softwareName models.SoftwareName
+	switch software := s.(type) {
+	case *agents.Mysqld:
+		softwareName = models.MysqldSoftwareName
+	case *agents.Xtrabackup:
+		softwareName = models.XtrabackupSoftwareName
+	case *agents.Xbcloud:
+		softwareName = models.XbcloudSoftwareName
+	case *agents.Qpress:
+		softwareName = models.QpressSoftwareName
+	default:
+		return "", errors.Errorf("invalid software type %T", software)
+	}
+
+	return softwareName, nil
+}
+
+// updateVersions updates software versions for one service.
+func (s *Service) updateVersions() (time.Duration, error) {
+	r, err := s.prepareUpdateVersions()
+	if err != nil {
+		return minCheckInterval, err
+	}
+
+	if !r.NeedsUpdate {
+		return r.CheckAfter, nil
+	}
+
+	softwares := []agents.Software{&agents.Mysqld{}, &agents.Xtrabackup{}, &agents.Xbcloud{}, &agents.Qpress{}}
+	versions, err := s.v.GetVersions(r.PMMAgentID, softwares)
+	if err != nil {
+		return minCheckInterval, err
+	}
+	if len(versions) != len(softwares) {
+		return minCheckInterval, errors.Errorf("slices length mismatch: versions len %d != softwares len %d",
+			len(versions), len(softwares))
+	}
```

Comment:
this check is not backward compatible, in case we add new softwares, but do not update client there always will be length mismatch.

Reply:
and the same in `GetVersions`

Reply:
returning filled `Error` field instead of the `Unknown` error: https://github.com/percona/pmm-agent/pull/283/commits/ecfa0dbc846ca5ff42d038823f385f37ba35e8e2